go/database/mpt/proof/verification_test.go

```diff
@@ -0,0 +1,672 @@
+	}
+	if len(add) != 1 {
+		t.Errorf("expected 1 address, got %d", len(add))
+	}
+}
+
+func TestVerification_Generates_ExistingKey(t *testing.T) {
+	ctrl := gomock.NewController(t)
+
+	trie := NewMockverifiableTrie(ctrl)
+
+	gomock.InOrder(
+		trie.EXPECT().GetValue(gomock.Any(), gomock.Any()).Return(common.Value{1}, nil),
+		trie.EXPECT().GetValue(gomock.Any(), gomock.Any()).Return(common.Value{}, nil),
+	)
+
+	keys, err := generateUnusedKeys(trie, 1, common.Address{})
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+	if len(keys) != 1 {
+		t.Errorf("expected 1 key, got %d", len(keys))
+	}
+}
+
+func TestVerification_Log_Processed_Accounts(t *testing.T) {
+	ctrl := gomock.NewController(t)
+
+	accountNode := &mpt.AccountNode{}
+	const LogWindow = 100
+
+	observer := mpt.NewMockVerificationObserver(ctrl)
+	observer.EXPECT().Progress(gomock.Any()).Do(func(msg string) {
+		if !strings.Contains(msg, fmt.Sprintf("  ... verified %d addresses", LogWindow)) {
+			t.Errorf("expected to see a log message with the number of processed accounts = %d", LogWindow)
+		}
+	})
+
+	proof := witness.NewMockProof(ctrl)
+	proof.EXPECT().GetBalance(gomock.Any(), gomock.Any()).Return(amount.New(), true, nil).AnyTimes()
+	proof.EXPECT().GetNonce(gomock.Any(), gomock.Any()).Return(common.Nonce{}, true, nil).AnyTimes()
+	proof.EXPECT().GetCodeHash(gomock.Any(), gomock.Any()).Return(common.Hash{}, true, nil).AnyTimes()
+	proof.EXPECT().GetState(gomock.Any(), gomock.Any(), gomock.Any()).Return(common.Value{}, true, nil).AnyTimes()
+
+	trie := NewMockverifiableTrie(ctrl)
+	trie.EXPECT().CreateWitnessProof(gomock.Any(), gomock.Any()).Return(proof, nil).AnyTimes()
+	trie.EXPECT().VisitAccount(gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+	trie.EXPECT().GetValue(gomock.Any(), gomock.Any()).Return(common.Value{}, nil).AnyTimes()
+
+	visitor := accountVerifyingVisitor{trie: trie, observer: observer, logWindow: LogWindow,
+		ctx: context.Background()}
+
+	for i := 0; i < LogWindow+1; i++ {
+		visitor.Visit(accountNode, mpt.NodeInfo{})
+	}
+}
+
+// countingWhenDoneContext is a context.Context that counts the number of times Done is called, and signals done only
+// when the threshold is reached.
+type countingWhenDoneContext struct {
+	context.Context
+	count     int // count the number of executions checking done
+	threshold int // when this threshold is reached, signal done from this point onwards
+	done      chan struct{}
+}
+
+func newCountingWhenDoneContext(ctx context.Context, threshold int) *countingWhenDoneContext {
+	return &countingWhenDoneContext{
+		Context:   ctx,
+		threshold: threshold,
+		done:      make(chan struct{}),
+	}
+}
+
+func (c *countingWhenDoneContext) Done() <-chan struct{} {
+	if c.count == c.threshold { // equality to close only once
+		close(c.done)
+	}
+	c.count++
+	return c.done
+}
+
+// errorInjectingProof is a witness.Proof that can be configured to return an error or wrong values after a certain number of calls.
```

Comment:
Could the same be done with less code using a mock of [Proof](https://github.com/Fantom-foundation/Carmen/blob/3fc9e177f6bb15f517db30573f9b50266706de04/go/common/witness/proof.go#L25)?

Reply:
yes, switched to use mocks. Not sure the result is very much better, but I think it is ok